feature-account-impl/src/main/java/io/novafoundation/nova/feature_account_impl/data/fee/types/HydrationConversionFeePayment.kt

```diff
@@ -0,0 +1,57 @@
+import io.novafoundation.nova.feature_account_api.data.model.Fee
+import io.novafoundation.nova.feature_account_api.data.model.SubstrateFee
+import io.novafoundation.nova.feature_account_api.domain.interfaces.AccountRepository
+import io.novafoundation.nova.feature_account_impl.data.fee.utils.HydraDxQuoteSharedComputation
+import io.novafoundation.nova.feature_swap_core.data.network.HydraDxAssetIdConverter
+import io.novafoundation.nova.feature_swap_core.data.network.setFeeCurrency
+import io.novafoundation.nova.feature_swap_core.data.network.toOnChainIdOrThrow
+import io.novafoundation.nova.runtime.ext.commissionAsset
+import io.novafoundation.nova.runtime.ext.fullId
+import io.novafoundation.nova.runtime.multiNetwork.ChainRegistry
+import io.novafoundation.nova.runtime.multiNetwork.asset
+import io.novafoundation.nova.runtime.multiNetwork.chain.model.Chain
+import io.novasama.substrate_sdk_android.runtime.extrinsic.ExtrinsicBuilder
+import kotlinx.coroutines.CoroutineScope
+
+internal class HydrationConversionFeePayment(
+    private val paymentAsset: Chain.Asset,
+    private val chainRegistry: ChainRegistry,
+    private val hydraDxAssetIdConverter: HydraDxAssetIdConverter,
+    private val hydraDxQuoteSharedComputation: HydraDxQuoteSharedComputation,
+    private val accountRepository: AccountRepository,
+    private val coroutineScope: CoroutineScope
+) : FeePayment {
+
+    override suspend fun modifyExtrinsic(extrinsicBuilder: ExtrinsicBuilder) {
+        val baseCall = extrinsicBuilder.getCall()
+        extrinsicBuilder.resetCalls()
+
+        extrinsicBuilder.setFeeCurrency(hydraDxAssetIdConverter.toOnChainIdOrThrow(paymentAsset))
+        extrinsicBuilder.call(baseCall)
+        extrinsicBuilder.setFeeCurrency(hydraDxAssetIdConverter.systemAssetId)
+    }
+
+    override suspend fun convertNativeFee(nativeFee: Fee): Fee {
+        val metaAccount = accountRepository.getSelectedMetaAccount()
+        val chain = chainRegistry.getChain(paymentAsset.chainId)
+        val accountId = metaAccount.accountIdIn(chain)
+        val fromAsset = chain.commissionAsset
+
+        val quote = hydraDxQuoteSharedComputation.quote(chain, accountId!!, fromAsset = fromAsset, toAsset = paymentAsset, coroutineScope)
+        return SubstrateFee(quote.quote, nativeFee.submissionOrigin, paymentAsset.fullId)
+    }
+
+    override suspend fun availableCustomFeeAssets(): List<Chain.Asset> {
+        val metaAccount = accountRepository.getSelectedMetaAccount()
+        val chain = chainRegistry.getChain(paymentAsset.chainId)
+        val accountId = metaAccount.accountIdIn(chain)
+        val fromAsset = chain.commissionAsset
+
+        val allSwapDirections = hydraDxQuoteSharedComputation.directions(chain, accountId!!, coroutineScope)
+        val commissionAssetDirections = allSwapDirections.adjacencyList[fromAsset.fullId] ?: emptyList()
+        return commissionAssetDirections.map { chainRegistry.asset(it.direction.to) }
```

Comment:
Are we sure this will handle a case when there is no direct route between HDX and fee token but there is an inderct route for it? E.g. HDX and USDT:
There is no direct HDX -> USDT route, but there is HDX -> StablePool -> USDT one

I cant recall whether adjencyList returned from hydra swap source includes only direct connections or all possible ones. I feel like the latter should be the case but its better to check